File: xpenbox-back/src/main/java/org/xpenbox/payment/dto/PreApprovalSubscriptionRequestDTO.java
package org.xpenbox.payment.dto;

import org.hibernate.validator.constraints.Length;
import org.xpenbox.payment.enums.PaymentProviderType;

import io.quarkus.runtime.annotations.RegisterForReflection;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

/**
 * Data Transfer Object for pre-approval plan request. This DTO is used to encapsulate the data required to create a pre-approval plan for a subscription. It includes validation annotations to ensure that the necessary fields are provided and meet certain criteria, such as not being null or blank and having a maximum length. The resource code of the plan is a crucial piece of information that identifies the specific plan for which the pre-approval is being requested.
 * @param resourceCodePlan the resource code of the plan for which the pre-approval is being requested, must not be null, blank, and must not exceed 100 characters in length
 * @param paymentProviderType the type of payment provider for the pre-approval plan, must not be null
 */
@RegisterForReflection
public record PreApprovalSubscriptionRequestDTO (

    @NotNull(message = "Resource code of the plan is required")
    @NotBlank(message = "Resource code of the plan must not be blank")
    @Length(max = 100, message = "Resource code must not exceed 100 characters")
    String resourceCodePlan,

    @NotNull(message = "Payment provider type is required")
    PaymentProviderType paymentProviderType
) { }

File: xpenbox-front/src/app/feature/subscription/model/pre-approval-subscription.request.dto.ts

/**
 * Data Transfer Object (DTO) for the request to create a pre-approval subscription.
 * @param resourceCodePlan - The code of the subscription plan to which the user wants to subscribe.
 * @param paymentProviderType - The type of payment provider to be used for the subscription.
 */
export class PreApprovalSubscriptionRequestDTO {
    resourceCodePlan: string;
    paymentProviderType: string;
    constructor(resourceCodePlan: string, paymentProviderType: string) {
        this.resourceCodePlan = resourceCodePlan;
        this.paymentProviderType = paymentProviderType;
    }

    /**
     * Generates a request body for creating a pre-approval subscription using predefined constants for the plan and provider.
     * @returns A new instance of PreApprovalSubscriptionRequestDTO with the specified plan and provider.
     */
    static generateRequestBody(): PreApprovalSubscriptionRequestDTO {
        return new PreApprovalSubscriptionRequestDTO(SubscriptionConstants.plan, SubscriptionConstants.provider);
    }
}

/**
 * Constants used for subscription requests, including the payment provider and the subscription plan.
 * provider - The payment provider to be used for the subscription (e.g., 'MERCADOPAGO').
 * plan - The code of the subscription plan to which the user wants to subscribe (e.g., 'plan_pro_monthly').
 */
export const SubscriptionConstants = {
    provider: 'MERCADOPAGO',
    plan: 'plan_pro_monthly'
}
